### PR TITLE
Remove AWSClient EventLoop APIs

### PR DESCRIPTION
**Pull request by adam-fowler**
- AWSClient.execute functions
- AWSClient.getCredential
- AWSClient.signURL/signHeaders
- Fix up waiter, paginate, service and STSAssume code reliant on EventLoop based AWSClient.execute
- Merge AWSClient, AWSService and AWSClient+Waiter files with async files

### CODE REVIEW

File: Sources/SotoCore/Credential/STSAssumeRole.swift [hunk 1]
```diff
@@ -85,11 +85,11 @@ struct STSCredentials: AWSDecodableShape, ExpiringCredential {
 }
 
 /// Credential Provider that holds an AWSClient
-protocol CredentialProviderWithClient: CredentialProvider {
+protocol AsyncCredentialProviderWithClient: AsyncCredentialProvider {
```
Comment (0xTim) [inline Sources/SotoCore/Credential/STSAssumeRole.swift:88]: Is it worth just renaming these to be `CredentialProvider` etc without the `Async-` prefix now that everything is async by default?

File: Sources/SotoCore/Credential/STSAssumeRole.swift [hunk 2]
```diff
@@ -131,17 +131,22 @@ struct STSAssumeRoleCredentialProvider: CredentialProviderWithClient {
...
 
-    func getCredential(on eventLoop: EventLoop, logger: Logger) -> EventLoopFuture<Credential> {
-        self.assumeRole(self.request, logger: logger, on: eventLoop)
-            .flatMapThrowing { response in
-                guard let credentials = response.credentials else {
-                    throw CredentialProviderError.noProvider
-                }
-                return credentials
-            }
+    func getCredential(on eventLoop: EventLoop, logger: Logger) async throws -> Credential {
```
Comment (0xTim) [inline Sources/SotoCore/Credential/STSAssumeRole.swift:134]: Do we need the event loop for this still?

File: Sources/SotoCore/Credential/STSAssumeRole.swift [hunk 3]
```diff
@@ -131,17 +131,22 @@ struct STSAssumeRoleCredentialProvider: CredentialProviderWithClient {
...
+        let response = try await self.assumeRole(self.request, logger: logger, on: eventLoop)
+        guard let credentials = response.credentials else {
+            throw CredentialProviderError.noProvider
+        }
+        return credentials
     }
 
-    func assumeRole(_ input: STSAssumeRoleRequest, logger: Logger, on eventLoop: EventLoop?) -> EventLoopFuture<STSAssumeRoleResponse> {
-        return self.client.execute(operation: "AssumeRole", path: "/", httpMethod: .POST, serviceConfig: self.config, input: input, logger: logger, on: eventLoop)
+    func assumeRole(_ input: STSAssumeRoleRequest, logger: Logger, on eventLoop: EventLoop?) async throws -> STSAssumeRoleResponse {
```
Comment (0xTim) [inline Sources/SotoCore/Credential/STSAssumeRole.swift:142]: You can remove the event loop as it's not used

File: Sources/SotoCore/Waiters/AWSClient+Waiter.swift [hunk 4]
```diff
@@ -91,4 +91,71 @@ extension AWSClient {
...
+    /// Returns when waiter polling returns a success state
+    /// or returns an error if the polling returns an error or timesout
+    ///
+    /// - Parameters:
+    ///   - input: Input parameters
+    ///   - waiter: Waiter to wait on
+    ///   - maxWaitTime: Maximum amount of time to wait
+    ///   - logger: Logger used to provide output
+    ///   - eventLoop: EventLoop to run API calls on
+    /// - Returns: EventLoopFuture that will be fulfilled once waiter has completed
```
Comment (0xTim) [inline Sources/SotoCore/Waiters/AWSClient+Waiter.swift:104]: Doesn't return anything anymore

File: Sources/SotoCore/Waiters/AWSClient+Waiter.swift [hunk 5]
```diff
@@ -91,4 +91,71 @@ extension AWSClient {
...
+    public func waitUntil<Input, Output>(
+        _ input: Input,
+        waiter: Waiter<Input, Output>,
+        maxWaitTime: TimeAmount? = nil,
+        logger: Logger = AWSClient.loggingDisabled,
+        on eventLoop: EventLoop? = nil
```
Comment (0xTim) [inline Sources/SotoCore/Waiters/AWSClient+Waiter.swift:110]: Event loop still being passed in

**Review by 0xTim — CHANGES_REQUESTED** [review]
There seem to be a few left over event loops in the PR

File: Sources/SotoCore/Credential/STSAssumeRole.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/SotoCore/Credential/STSAssumeRole.swift:88]: When I get to making all the CredentialProviders async I will do this.

**Review by 0xTim — APPROVED** [review]

File: Sources/SotoCore/Waiters/AWSClient+Waiter.swift [hunk 6]
```diff
@@ -91,4 +91,69 @@ extension AWSClient {
...
+
+    /// Returns when waiter polling returns a success state
+    /// or returns an error if the polling returns an error or timesout
+    ///
+    /// - Parameters:
+    ///   - input: Input parameters
+    ///   - waiter: Waiter to wait on
+    ///   - maxWaitTime: Maximum amount of time to wait
+    ///   - logger: Logger used to provide output
+    ///   - eventLoop: EventLoop to run API calls on
```
Comment (0xTim) [inline Sources/SotoCore/Waiters/AWSClient+Waiter.swift:103]:
```suggestion
    ///   - logger: Logger used to provide output
```

**Pull request merged**